README
draw.io project sources
-----------------------

draw.io comprises three parts:

1) The static web content under the war directory.
2) The open, save, embed, proxy and iconfinder servlets as one servlet war. These are located under src/com/mxgraph/online/.
3) The image export servlet. This is located under etc/imageexport.

Prior to invoking any builds, search war/index.html for "CUSTOM_PARAMETERS". The image export URL is the most important variable to change (EXPORT_URL). 
Without that the export to image will use our server, which somewhat defeats the point of hosting it. The servlet is mapped to "export" and the default name 
of the war is "Imageexport" so "/Imageexport/export" is the value if you deploy that war as-is and the servlet engine is on the root of the domain.

/etc/build/build.xml is the ant build file for the main war. Invoking "war" will generate build/draw.war.

/etc/imagexport/build.xml is the ant build file for the image export war. Invoking "war" will generate /etc/imageexport/imageexport.war.

Depoly these wars on a servlet engine and navigate to the location of the first war, you will get the draw.io application.

Licensing
---------

The underlying mxGraph library is licensed to you either the http://creativecommons.org/licenses/by-nc-sa/3.0/ or the mxGraph commercial license 
(if you're a paying customer). Please understand that we make a living selling the mxGraph library commercially, we need to avoid draw.io being 
used freely where mxGraph would previously have been sold.

The source to draw.io itself is Apache 2.0 licensed (or mxGraph Commercial License), but obviously the mxGraph license dominates this for commercial 
purposes.

We separate between developing against the mxGraph API and just deploying draw.io for general usage. Thus, we make exceptions to the CCNC license 
if you want to take draw.io as-is and deploy it for your own use (note this only affects commercial users). Below, we will build up a list of alterations 
you can make to draw.io without being subject to the mxGraph commercial license (note this only affects commercial users):

- You can alter draw.io paths so that the functionality works correctly in your deployment environment
- You can add and remove menu options and library stencil sets to/from the UI

If you require more exceptions, please add to issue tracker.

ChangeLog
05-FEB-2013: 2.1.0.1

- Fixes 7 second repeating Google auth dialog bug

04-FEB-2013: 2.1.0.0

- Removes old sharing
- Fixes relative images in embedded diagrams
- Uses App Engne SDK 1.7.4
- Uses mxGraph 2.0.0.0

16-JAN-2013: 2.0.1.7

- Uses mxGraph 1.10.4.2
- Adds curved format menu item
- Uses extended image picker if logged into Google
- Removes devhost URL parameter

20-DEC-2012: 2.0.1.6

- Fixes issue with Google Drive files retaining file name

19-DEC-2012: 2.0.1.5

- Enabled .vdx open/importing (alpha)

19-DEC-2012: 2.0.1.4

- Shows spinner while loading url URL parameter
- Fixes alignment of SVG sidebar entries in IE
- Fixes flickering of sidebar entries in FF
- Adds spinner to minified diagramly

18-DEC-2012: 2.0.1.3

- Uses mxGraph 1.10.4.1
- Fixes handling of url URL parameter

14-DEC-2012: 2.0.1.2

- Removes plugins feature

13-DEC-2012: 2.0.1.1

- Adds security alert when loading plugins
- Fixes clipping bug

12-DEC-2012: 2.0.1.0

- Fixes session expiry issue in Google Drive

12-DEC-2012: 2.0.0.4

- Adds plugins via URL parameter
- Fixes toolbar size
- Fixes browser context menu
- Adds colors in color dialog

06-DEC-2012: 2.0.0.1

- Switches to using JS client library for Google Drive calls

08-NOV-2012: 1.6.10.0

- Adds Korean, Japanese, Russian, Bosnian and Ukrainian translations
- Switches to port 80 for image export server

31-OCT-2012: 1.6.9.2

- Removed data nucleus libraries
- Adds rawdeflate, base64 to diagramly.min.js
- Changes color of social icons to gray

30-OCT-2012: 1.6.9.1

- Removes social JS
- Uses mxGraph 1.10.4.0

25-OCT-2012: 1.6.9.0

- Adds Google Drive install scope
- Updates libraries used for Google Drive integration

19-OCT-2012: 1.6.8.1

- Fixes handling of file positioning in Drive

16-OCT-2012: 1.6.8.0

- Fixes using multiple accounts with Google Drive

24-SEP-2012: 1.6.7.4

- Changes diagram.ly to draw.io in dialogs
- Minor formatting fixes

21-SEP-2012: 1.6.7.3

- Add session tidying cron job for GAE

14-SEP-2012: 1.6.7.2

- Removes Facebook footer in IE6
- Adds layout options via grapheditor example

13-SEP-2012: 1.6.7.1

- Fixes use of custom handlers with closure compiler

31-AUG-2012: 1.6.7.0

- Uses mxGraph 1.10.3.1
- Fixes background color for SVG export in quirks mode
- Enables Google image picker on Mac/iPad
- Adds Github menu item in help menu

29-AUG-2012: 1.6.1.9

- Fixes Github deployment Ant task
- Fixes German Facebook button (changes to English)
- Fixes touch-based panning for embedded diagrams
- Uses mxGraph 1.10.3.0

22-AUG-2012: 1.6.1.8

- Adds Facebook like button

21-AUG-2012: 1.6.1.7

- Adds compression for embedded diagrams

02-AUG-2012: 1.6.1.6

- Adds fixes for null user objects in link handling

02-AUG-2012: 1.6.1.5

- Adds transparency option for PNG image, SVG export

01-AUG-2012: 1.6.1.4

- Fixes for filepicker.io

31-JUL-2012: 1.6.1.3

- Uses mxClient.js, mxgraph-core.jar 1.10.2.1
- Adds social, analytics URL parameters
- Adds persistent stats for embedding
- Adds save for filepicker.io

28-JUL-2012: 1.6.1.2

- Fixes embedding of international characters and newlines
- Changes image export URL to http://shr.diagramly.com:8080/ImageExport/export
- Changes lastModified header creation in EmbedServlet
- Fixes infinite recursion in Embed.js for convertValueToString

27-JUL-2012: 1.6.1.1

- Adds support for hyperlinks

23-JUL-2012: 1.6.1.0

- Uses mxGraph 1.10.2.0
- Adds embed feature
- Fixes errors if tinyMCE not available
- Fixes syntax of dash patterns in shapes

30-JUN-2012: 1.6.0.2

- Renames Init.js to Devel.js
- Adds link to "Did you know..." topic in help menu
- Removes previous diagramly link from help menu
- Adds url URL parameter handling in client
- Adds proxy servlet for loading files from URLs

22-JUN-2012: 1.6.0.1

- Fixes minor bugs
- Uses mxGraph 1.10.1.3

19-JUN-2012: 1.6.0.0

- Adds real-time collaboration (experimental)
- Adds handling for App not installed error in Google Drive
- Uses mxGraph 1.10.1.2

06-JUN-2012: 1.5.0.7

- Uses mxGraph 1.10.1.1
- Adds workaround for focus repaint bug in IE9/10
- Adds fixed connection points for custom shapes
- Adds closure compiler pre-deploy step
- Puts all JS in a single minified file

25-MAY-2012: 1.5.0.6

- Uses static mxGraph 1.10.0.6
- Fixes "Save as" in Google Drive
- Fixes page view, default pageScale
- Adds new items in file menu
- Adds options in file edit dialog
- Adds import action
- Adds extra handle for custom shapes

17-MAY-2012: 1.5.0.5

- Removes .mxe default file extension
- Changes mime type to application/mxe

16-MAY-2012: 1.5.0.4

- Uses current protocol to load picker API
- Adds CSS3 transitions for hover styles
- Simplifies tinyMCE toolbar, uses main toolbar (non-IE only)
- Adds "Use as default edge" action

14-MAY-2012: 1.5.0.3

- Adds border in export dialog
- Adds resources for Google Drive integration
- Fixes use of optional flash for saving in savedialog

05-MAY-2012: 1.5.0.2

- Handles requests for non-existent script versions
- Handles DeadlineExceededException

03-MAY-2012: 1.5.0.1

- Adds XML export option
- Fixes unicode character encoding in files
- Adds status information for Google Drive operations
- Adds spinner during Google Drive operations
- Adds reconnect for lost session after save
- Fixes illegal state exceptions in servlet

02-MAY-2012: 1.5.0.0

- Adds working Google Drive integration


war/js/mxgraph/Actions.js
/**
 * $Id: Actions.js,v 1.6 2013-01-17 13:42:26 gaudenz Exp $
 * Copyright (c) 2006-2012, JGraph Ltd
 */
/**
 * Constructs the actions object for the given UI.
 */
function Actions(editorUi)
{
	this.editorUi = editorUi;
	this.actions = new Object();
	this.init();
};

/**
 * Adds the default actions.
 */
Actions.prototype.init = function()
{
	var ui = this.editorUi;
	var editor = ui.editor;
	var graph = editor.graph;

	// File actions
	this.addAction('new', function() { window.open(ui.getUrl()); });
	this.addAction('open', function()
	{
		window.openNew = true;
		window.openKey = 'open';
		
		ui.openFile();
	});
	this.addAction('import', function()
	{
		window.openNew = false;
		window.openKey = 'import';
		
		// Closes dialog after open
		window.openFile = new OpenFile(mxUtils.bind(this, function()
		{
			ui.hideDialog();
		}));
		
		window.openFile.setConsumer(mxUtils.bind(this, function(xml, filename)
		{
			try
			{
				var doc = mxUtils.parseXml(xml);
				var model = new mxGraphModel();
				var codec = new mxCodec(doc);
				codec.decode(doc.documentElement, model);
				
				var children = model.getChildren(model.getChildAt(model.getRoot(), 0));
				editor.graph.setSelectionCells(editor.graph.importCells(children));
			}
			catch (e)
			{
				mxUtils.alert(mxResources.get('invalidOrMissingFile') + ': ' + e.message);
			}
		}));

		// Removes openFile if dialog is closed
		ui.showDialog(new OpenDialog(this).container, 300, 180, true, true, function()
		{
			window.openFile = null;
		});
	});
	this.addAction('save', function() { ui.saveFile(false); }, null, null, 'Ctrl+S');
	this.addAction('saveAs', function() { ui.saveFile(true); }, null, null, 'Ctrl+Shift-S');
	this.addAction('export', function() { ui.showDialog(new ExportDialog(ui).container, 300, 200, true, true); }, null, null, 'Ctrl+E');
	this.put('editFile', new Action(mxResources.get('edit'), mxUtils.bind(this, function()
	{
		this.editorUi.showDialog(new EditFileDialog(ui).container, 620, 420, true, true);
	})));
	this.addAction('pageSetup', function() { ui.showDialog(new PageSetupDialog(ui).container, 300, 200, true, true); });
	this.addAction('print', function() { ui.showDialog(new PrintDialog(ui).container, 300, 200, true, true); }, null, 'sprite-print', 'Ctrl+P');
	this.addAction('preview', function() { mxUtils.show(graph, null, 10, 10); });
	
	// Edit actions
	this.addAction('undo', function() { editor.undoManager.undo(); }, null, 'sprite-undo', 'Ctrl+Z');
	this.addAction('redo', function() { editor.undoManager.redo(); }, null, 'sprite-redo', 'Ctrl+Y');
	this.addAction('cut', function() { mxClipboard.cut(graph); }, null, 'sprite-cut', 'Ctrl+X');
	this.addAction('copy', function() { mxClipboard.copy(graph); }, null, 'sprite-copy', 'Ctrl+C');
	this.addAction('paste', function() { mxClipboard.paste(graph); }, false, 'sprite-paste', 'Ctrl+V');
	this.addAction('delete', function() { graph.removeCells(); }, null, null, 'Delete');
	this.addAction('duplicate', function()
    {
		var s = graph.gridSize;
		graph.setSelectionCells(graph.moveCells(graph.getSelectionCells(), s, s, true));
    }, null, null, 'Ctrl+D');
	this.addAction('selectVertices', function() { graph.selectVertices(); }, null, null, 'Ctrl+Shift+V');
	this.addAction('selectEdges', function() { graph.selectEdges(); }, null, null, 'Ctrl+Shift+E');
	this.addAction('selectAll', function() { graph.selectAll(); }, null, null, 'Ctrl+A');

	// Navigation actions
	this.addAction('home', function() { graph.home(); }, null, null, 'Home');
	this.addAction('exitGroup', function() { graph.exitGroup(); }, null, null, 'Page Up');
	this.addAction('enterGroup', function() { graph.enterGroup(); }, null, null, 'Page Down');
	this.addAction('expand', function() { graph.foldCells(false); }, null, null, 'Enter');
	this.addAction('collapse', function() { graph.foldCells(true); }, null, null, 'Backspace');

	// Arrange actions
	this.addAction('toFront', function() { graph.orderCells(false); }, null, null, 'Ctrl+F');
	this.addAction('toBack', function() { graph.orderCells(true); }, null, null, 'Ctrl+B');
	this.addAction('group', function() { graph.setSelectionCell(graph.groupCells(null, 0)); }, null, null, 'Ctrl+G');
	this.addAction('ungroup', function() { graph.setSelectionCells(graph.ungroupCells()); }, null, null, 'Ctrl+U');
	this.addAction('removeFromGroup', function() { graph.removeCellsFromParent(); });
	this.addAction('editLink', function()
	{
		var cell = graph.getSelectionCell();
		var link = graph.getLinkForCell(cell);
		
		if (link == null)
		{
			link = '';
		}
		
		link = mxUtils.prompt(mxResources.get('enterValue'), link);
		
		if (link != null)
		{
			graph.setLinkForCell(cell, link);
		}
	});
	this.addAction('openLink', function()
	{
		var cell = graph.getSelectionCell();
		var link = graph.getLinkForCell(cell);
		
		if (link != null)
		{
			window.open(link);
		}
	});
	this.addAction('autosize', function()
	{
		var cells = graph.getSelectionCells();
		
		if (cells != null)
		{
			graph.getModel().beginUpdate();
			try
			{
				for (var i = 0; i < cells.length; i++)
				{
					var cell = cells[i];
					
					if (graph.getModel().getChildCount(cell))
					{
						graph.updateGroupBounds([cell], 20);
					}
					else
					{
						graph.updateCellSize(cell);
					}
				}
			}
			finally
			{
				graph.getModel().endUpdate();
			}
		}
	});
	this.addAction('wordWrap', function()
	{
    	var state = graph.getView().getState(graph.getSelectionCell());
    	var value = 'wrap';
    	
    	if (state != null && state.style[mxConstants.STYLE_WHITE_SPACE] == 'wrap')
    	{
    		value = null;
    	}

       	graph.setCellStyles(mxConstants.STYLE_WHITE_SPACE, value);
	});
	this.addAction('rotation', function()
	{
		var value = '0';
    	var state = graph.getView().getState(graph.getSelectionCell());
    	
    	if (state != null)
    	{
    		value = state.style[mxConstants.STYLE_ROTATION] || value;
    	}

		value = mxUtils.prompt(mxResources.get('enterValue') + ' (' +
				mxResources.get('rotation') + ' 0-360)', value);
        	
    	if (value != null)
    	{
        	graph.setCellStyles(mxConstants.STYLE_ROTATION, value);
        }
	});
	this.addAction('tilt', function()
	{
		var cells = graph.getSelectionCells();
		
		if (cells != null)
		{
			graph.getModel().beginUpdate();
			try
			{
				for (var i = 0; i < cells.length; i++)
				{
					var cell = cells[i];
					
					if (graph.getModel().isVertex(cell) && graph.getModel().getChildCount(cell) == 0)
					{
						var geo = graph.getCellGeometry(cell);
			
						if (geo != null)
						{
							// Rotates the size and position in the geometry
							geo = geo.clone();
							geo.x += geo.width / 2 - geo.height / 2;
							geo.y += geo.height / 2 - geo.width / 2;
							var tmp = geo.width;
							geo.width = geo.height;
							geo.height = tmp;
							graph.getModel().setGeometry(cell, geo);
							
							// Reads the current direction and advances by 90 degrees
							var state = graph.view.getState(cell);
							
							if (state != null)
							{
								var dir = state.style[mxConstants.STYLE_DIRECTION] || 'east'/*default*/;
								
								if (dir == 'east')
								{
									dir = 'south';
								}
								else if (dir == 'south')
								{
									dir = 'west';
								}
								else if (dir == 'west')
								{
									dir = 'north';
								}
								else if (dir == 'north')
								{
									dir = 'east';
								}
								
								graph.setCellStyles(mxConstants.STYLE_DIRECTION, dir, [cell]);
							}
						}
					}
				}
			}
			finally
			{
				graph.getModel().endUpdate();
			}
		}
	}, null, null, 'Ctrl+R');
	
	// View actions
	this.addAction('actualSize', function()
	{
		graph.zoomTo(1);
	});
	this.addAction('zoomIn', function() { graph.zoomIn(); }, null, null, 'Add');
	this.addAction('zoomOut', function() { graph.zoomOut(); }, null, null, 'Subtract');
	this.addAction('fitWindow', function() { graph.fit(); });

	this.addAction('fitPage', mxUtils.bind(this, function()
	{
		if (!graph.pageVisible)
		{
			this.get('pageView').funct();
		}
		
		var fmt = graph.pageFormat;
		var ps = graph.pageScale;
		var cw = graph.container.clientWidth - 20;
		var ch = graph.container.clientHeight - 20;
		
		var scale = Math.floor(100 * Math.min(cw / fmt.width / ps, ch / fmt.height / ps)) / 100;
		graph.zoomTo(scale);
		
		graph.container.scrollLeft = Math.round(graph.view.translate.x * scale - Math.max(10, (graph.container.clientWidth - fmt.width * ps * scale) / 2));
		graph.container.scrollTop = Math.round(graph.view.translate.y * scale - Math.max(10, (graph.container.clientHeight - fmt.height * ps * scale) / 2));
	}));
	this.addAction('fitPageWidth', mxUtils.bind(this, function()
	{
		if (!graph.pageVisible)
		{
			this.get('pageView').funct();
		}
		
		var fmt = graph.pageFormat;
		var ps = graph.pageScale;
		var cw = graph.container.clientWidth - 20;
		
		var scale = Math.floor(100 * cw / fmt.width / ps) / 100;
		graph.zoomTo(scale);
		
		graph.container.scrollLeft = Math.round(graph.view.translate.x * scale - Math.max(10, (graph.container.clientWidth - fmt.width * ps * scale) / 2));
		graph.container.scrollTop = Math.round(graph.view.translate.y * scale - Math.max(10, (graph.container.clientHeight - fmt.height * ps * scale) / 2));
	}));
	this.put('customZoom', new Action(mxResources.get('custom'), function()
	{
    	var value = mxUtils.prompt(mxResources.get('enterValue') + ' (%)', parseInt(graph.getView().getScale() * 100));
    	
    	if (value != null && value.length > 0 && !isNaN(parseInt(value)))
    	{
    		graph.zoomTo(parseInt(value) / 100);
        }
	}));
	
	// Option actions
	var action = null;
	action = this.addAction('grid', function()
	{
		graph.setGridEnabled(!graph.isGridEnabled());
		editor.updateGraphComponents();
	}, null, null, 'Ctrl+Shift+G');
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.isGridEnabled(); });
	action = this.addAction('guides', function() { graph.graphHandler.guidesEnabled = !graph.graphHandler.guidesEnabled; });
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.graphHandler.guidesEnabled; });
	action = this.addAction('tooltips', function()
	{
		graph.tooltipHandler.setEnabled(!graph.tooltipHandler.isEnabled());
	});
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.tooltipHandler.isEnabled(); });
	action = this.addAction('navigation', function()
	{
		graph.foldingEnabled = !graph.foldingEnabled;
    	graph.view.revalidate();
	});
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.foldingEnabled; });
	action = this.addAction('scrollbars', function()
	{
		graph.scrollbars = !graph.scrollbars;
		editor.updateGraphComponents();

		if (!graph.scrollbars)
		{
			var t = graph.view.translate;
			graph.view.setTranslate(t.x - graph.container.scrollLeft / graph.view.scale, t.y - graph.container.scrollTop / graph.view.scale);
			graph.container.scrollLeft = 0;
			graph.container.scrollTop = 0;
			graph.sizeDidChange();
		}
		else
		{
			var dx = graph.view.translate.x;
			var dy = graph.view.translate.y;

			graph.view.translate.x = 0;
			graph.view.translate.y = 0;
			graph.sizeDidChange();
			graph.container.scrollLeft -= Math.round(dx * graph.view.scale);
			graph.container.scrollTop -= Math.round(dy * graph.view.scale);
		}
	}, !mxClient.IS_TOUCH);
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.container.style.overflow == 'auto'; });
	action = this.addAction('pageView', mxUtils.bind(this, function()
	{
		graph.pageVisible = !graph.pageVisible;
		graph.pageBreaksVisible = graph.pageVisible; 
		graph.preferPageSize = graph.pageBreaksVisible;
		graph.view.validate();
		graph.sizeDidChange();
		
		editor.updateGraphComponents();
		editor.outline.update();
		
		if (mxUtils.hasScrollbars(graph.container))
		{
			if (graph.pageVisible)
			{
				graph.container.scrollLeft -= 20;
				graph.container.scrollTop -= 20;
			}
			else
			{
				graph.container.scrollLeft += 20;
				graph.container.scrollTop += 20;
			}
		}
	}));
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.pageVisible; });
	this.put('pageBackgroundColor', new Action(mxResources.get('backgroundColor'), function()
	{
		var apply = function(color)
		{
			graph.background = color;
			editor.updateGraphComponents();
		};

		var cd = new ColorDialog(ui, graph.background || 'none', apply);
		ui.showDialog(cd.container, 220, 400, true, false);
		
		if (!mxClient.IS_TOUCH)
		{
			cd.colorInput.focus();
		}
	}));
	action = this.addAction('connect', function()
	{
		graph.setConnectable(!graph.connectionHandler.isEnabled());
	}, null, null, 'Ctrl+Q');
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.connectionHandler.isEnabled(); });
	action = this.addAction('copyConnect', function()
	{
		graph.connectionHandler.setCreateTarget(!graph.connectionHandler.isCreateTarget());
	});
	action.setToggleAction(true);
	action.setSelectedCallback(function() { return graph.connectionHandler.isCreateTarget(); });
	
	// Help actions
	this.addAction('help', function()
	{
		var ext = '';
		
		if (mxResources.isLanguageSupported(mxClient.language))
		{
			ext = '_' + mxClient.language;
		}
		
		window.open(RESOURCES_PATH + '/help' + ext + '.html');
	});
	this.put('about', new Action(mxResources.get('about') + ' Graph Editor', function()
	{
		ui.showDialog(new AboutDialog(ui).container, 320, 280, true, true);
	}, null, null, 'F1'));
	
	// Font style actions
	var toggleFontStyle = mxUtils.bind(this, function(key, style)
	{
		this.addAction(key, function()
		{
			graph.toggleCellStyleFlags(mxConstants.STYLE_FONTSTYLE, style);
		});
	});
	
	toggleFontStyle('bold', mxConstants.FONT_BOLD);
	toggleFontStyle('italic', mxConstants.FONT_ITALIC);
	toggleFontStyle('underline', mxConstants.FONT_UNDERLINE);
	
	// Color actions
	this.addAction('fontColor', function() { ui.menus.pickColor(mxConstants.STYLE_FONTCOLOR); });
	this.addAction('strokeColor', function() { ui.menus.pickColor(mxConstants.STYLE_STROKECOLOR); });
	this.addAction('fillColor', function() { ui.menus.pickColor(mxConstants.STYLE_FILLCOLOR); });
	this.addAction('gradientColor', function() { ui.menus.pickColor(mxConstants.STYLE_GRADIENTCOLOR); });
	this.addAction('backgroundColor', function() { ui.menus.pickColor(mxConstants.STYLE_LABEL_BACKGROUNDCOLOR); });
	this.addAction('borderColor', function() { ui.menus.pickColor(mxConstants.STYLE_LABEL_BORDERCOLOR); });
	
	// Format actions
	this.addAction('shadow', function() { graph.toggleCellStyles(mxConstants.STYLE_SHADOW); });
	this.addAction('dashed', function() { graph.toggleCellStyles(mxConstants.STYLE_DASHED); });
	this.addAction('rounded', function() { graph.toggleCellStyles(mxConstants.STYLE_ROUNDED); });
	this.addAction('curved', function() { graph.toggleCellStyles(mxConstants.STYLE_CURVED); });
	this.addAction('style', function()
	{
		var cells = graph.getSelectionCells();
		
		if (cells != null && cells.length > 0)
		{
			var model = graph.getModel();
			var style = mxUtils.prompt(mxResources.get('enterValue')+ ' (' + mxResources.get('style') + ')',
					model.getStyle(cells[0]) || '');

			if (style != null)
			{
				graph.setCellStyle(style, cells);
			}
		}
	});
	this.addAction('setAsDefaultEdge', function()
	{
		graph.setDefaultEdge(graph.getSelectionCell());
	});
	this.addAction('addWaypoint', function()
	{
		var cell = graph.getSelectionCell();
		
		if (cell != null && graph.getModel().isEdge(cell))
		{
			var handler = editor.graph.selectionCellsHandler.getHandler(cell);
			
			if (handler instanceof mxEdgeHandler)
			{
				var t = graph.view.translate;
				handler.addPointAt(handler.state, graph.panningHandler.triggerX - t.x, graph.panningHandler.triggerY - t.y);
			}
		}
	});
	this.addAction('removeWaypoint', function()
	{
		// TODO: Action should run with "this" set to action
		var rmWaypointAction = ui.actions.get('removeWaypoint');
		
		if (rmWaypointAction.handler != null)
		{
			// NOTE: Popupevent handled and action updated in Menus.createPopupMenu
			rmWaypointAction.handler.removePoint(rmWaypointAction.handler.state, rmWaypointAction.index);
		}
	});
	this.addAction('image', function()
	{
		function updateImage(value, w, h)
		{
			var select = null;
			var cells = graph.getSelectionCells();
			
			graph.getModel().beginUpdate();
        	try
        	{
        		// Inserts new cell if no cell is selected
    			if (cells.length == 0)
    			{
    				var gs = graph.getGridSize();
    				cells = [graph.insertVertex(graph.getDefaultParent(), null, '', gs, gs, w, h)];
    				select = cells;
    			}
    			
        		graph.setCellStyles(mxConstants.STYLE_IMAGE, value, cells);
	        	graph.setCellStyles(mxConstants.STYLE_SHAPE, 'image', cells);
	        	
	        	if (graph.getSelectionCount() == 1)
	        	{
		        	if (w != null && h != null)
		        	{
		        		var cell = cells[0];
		        		var geo = graph.getModel().getGeometry(cell);
		        		
		        		if (geo != null)
		        		{
		        			geo = geo.clone();
			        		geo.width = w;
			        		geo.height = h;
			        		graph.getModel().setGeometry(cell, geo);
		        		}
		        	}
	        	}
        	}
        	finally
        	{
        		graph.getModel().endUpdate();
        	}
        	
        	if (select != null)
        	{
        		graph.setSelectionCells(select);
        		graph.scrollCellToVisible(select[0]);
        	}
		};

    	var value = '';
    	var state = graph.getView().getState(graph.getSelectionCell());
    	
    	if (state != null)
    	{
    		value = state.style[mxConstants.STYLE_IMAGE] || value;
    	}

    	value = mxUtils.prompt(mxResources.get('enterValue') + ' (' + mxResources.get('url') + ')', value);

    	if (value != null)
    	{
    		if (value.length > 0)
    		{
	    		var img = new Image();
	    		
	    		img.onload = function()
	    		{
	    			updateImage(value, img.width, img.height);
	    		};
	    		img.onerror = function()
	    		{
	    			mxUtils.alert(mxResources.get('fileNotFound'));
	    		};
	    		
	    		img.src = value;
    		}
        }
	});
};

/**
 * Registers the given action under the given name.
 */
Actions.prototype.addAction = function(key, funct, enabled, iconCls, shortcut)
{
	return this.put(key, new Action(mxResources.get(key), funct, enabled, iconCls, shortcut));
};

/**
 * Registers the given action under the given name.
 */
Actions.prototype.put = function(name, action)
{
	this.actions[name] = action;
	
	return action;
};

/**
 * Returns the action for the given name or null if no such action exists.
 */
Actions.prototype.get = function(name)
{
	return this.actions[name];
};

/**
 * Constructs a new action for the given parameters.
 */
function Action(label, funct, enabled, iconCls, shortcut)
{
	mxEventSource.call(this);
	this.label = label;
	this.funct = funct;
	this.enabled = (enabled != null) ? enabled : true;
	this.iconCls = iconCls;
	this.shortcut = shortcut;
};

// Action inherits from mxEventSource
mxUtils.extend(Action, mxEventSource);

/**
 * Sets the enabled state of the action and fires a stateChanged event.
 */
Action.prototype.setEnabled = function(value)
{
	if (this.enabled != value)
	{
		this.enabled = value;
		this.fireEvent(new mxEventObject('stateChanged'));
	}
};

/**
 * Sets the enabled state of the action and fires a stateChanged event.
 */
Action.prototype.setToggleAction = function(value)
{
	this.toggleAction = value;
};

/**
 * Sets the enabled state of the action and fires a stateChanged event.
 */
Action.prototype.setSelectedCallback = function(funct)
{
	this.selectedCallback = funct;
};

/**
 * Sets the enabled state of the action and fires a stateChanged event.
 */
Action.prototype.isSelected = function()
{
	return this.selectedCallback();
};
